syntax = "proto3";
package protocol;

enum Direction {
  FORWARD = 0;
  BACKWARD = 1;
}

enum LayerType {
  // Colors
  SingleColor = 0;
  RainbowColor = 1;
  SectionsWaveColor = 2;
  SectionsColor = 3;
  FadeColor = 4;
  SwitchColor = 5;

  // Masks - 50 offset
  BlinkMask = 50;
  InvertMask = 51;
  PulseSawtoothMask = 52;
  PulseMask = 53;
  SawtoothMask = 54;
  SectionsWaveMask = 55;
  SectionsMask = 56;
  StarsMask = 57;
  WaveMask = 58;
}

// The request message containing the desired effect and brightness.
message Layer {
  LayerType type = 1;
  uint32 duration = 2;
  uint32 length = 3;
  uint32 color = 4;
  uint32 gap = 5;
  uint32 frequency = 6;
  uint32 speed = 7;
  repeated uint32 colors = 8;
  bytes sections = 9;
}

// The request message containing an array of effects.
message Animation {
  Direction direction = 1;
  uint32 duration = 2;
  repeated Layer layers = 3;
}

message Sequence {
  uint32 brightness = 1;
  repeated Animation animations = 2;
}
